Remove use of .data in optimizers

albanD opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #30338 remove easy .data in torch/nn
* #30337 remote .dat from docs in torch/nn
* #30336 remove .data for torch/jit
* #30335 remove use of .data in torch/autograd
* #30334 remove dead python code that uses .data
* **#30258 Remove use of .data in optimizers**
* #30257 Make all optimizers consistent so that they don't change gradients inplace

Differential Revision: [D18665459](https://our.internmc.facebook.com/intern/diff/D18665459)

albanD committed:
Remove use of .data in optimizers

[ghstack-poisoned]

albanD committed:
Update on "Remove use of .data in optimizers"




[ghstack-poisoned]

albanD committed:
Update on "Remove use of .data in optimizers"




[ghstack-poisoned]

albanD committed:
Update on "Remove use of .data in optimizers"




[ghstack-poisoned]

vincentqb commented on torch/optim/adadelta.py:
```diff
@@ -35,6 +35,7 @@ def __init__(self, params, lr=1.0, rho=0.9, eps=1e-6, weight_decay=0):
         defaults = dict(lr=lr, rho=rho, eps=eps, weight_decay=weight_decay)
         super(Adadelta, self).__init__(params, defaults)
 
+    @torch.no_grad()
```
This is a clean way of disabling gradient computation within the stepping of the optimizers. It would be nice to be closer to enabling derivatives w/r to parameters of the optimizer. Offline you suggested replacing `@torch.no_grad()` with `with torch.no_grad():` inside the step function. Let's keep it this way for now. It might be easier to do while refactoring anyway, e.g. as part of #30232.

albanD replied:
We need to have a more high level thinking if we want to enable derivatives wrt the steps.
- What will be the api
- How will the user use this?
- Are the operations in all optimizers actually differentiable (lbfgs for example)

Definitely not in the scope of this PR.

vincentqb replied:
Agree.

vincentqb reviewed (approved):
LGTM

albanD committed:
Update on "Remove use of .data in optimizers"


Differential Revision: [D18665459](https://our.internmc.facebook.com/intern/diff/D18665459)

[ghstack-poisoned]

vincentqb commented:
Is this test related? It doesn't to me.

```
Nov 23 00:24:20 ERROR [0.077s]: test_tensor_sharing (jit.test_data_parallel.TestDataParallel)
Nov 23 00:24:20 ----------------------------------------------------------------------
Nov 23 00:24:20 Traceback (most recent call last):
Nov 23 00:24:20   File "/var/lib/jenkins/workspace/test/jit/test_data_parallel.py", line 118, in test_tensor_sharing
Nov 23 00:24:20     r0_forward = replica[0].forward(x)
Nov 23 00:24:20   File "/var/lib/jenkins/workspace/test/common_utils.py", line 82, in prof_meth_call
Nov 23 00:24:20     return prof_callable(meth_call, *args, **kwargs)
Nov 23 00:24:20   File "/var/lib/jenkins/workspace/test/common_utils.py", line 76, in prof_callable
Nov 23 00:24:20     return callable(*args, **kwargs)
Nov 23 00:24:20 RuntimeError: diff_view_meta->output_nr_ == 0 INTERNAL ASSERT FAILED at /var/lib/jenkins/workspace/torch/csrc/autograd/variable.cpp:326, please report a bug to PyTorch. 
Nov 23 00:24:20 The above operation failed in interpreter.
Nov 23 00:24:20 Traceback (most recent call last):
Nov 23 00:24:20   File "/opt/conda/lib/python3.6/site-packages/torch/nn/functional.py", line 1370
Nov 23 00:24:20     if input.dim() == 2 and bias is not None:
Nov 23 00:24:20         # fused op is marginally faster
Nov 23 00:24:20         ret = torch.addmm(bias, input, weight.t())
Nov 23 00:24:20               ~~~~~~~~~~~ <--- HERE
Nov 23 00:24:20     else:
Nov 23 00:24:20         output = input.matmul(weight.t())
```

vincentqb commented on torch/__init__.py:
```diff
@@ -272,14 +272,25 @@ def manager_path():
+
+# Warning: torch.no_grad cannot be used for module-level functions in
+# torch.autograd as it is only available after this line
+from torch.autograd import no_grad, enable_grad, set_grad_enabled
```
Thanks for updating this.

albanD commented:
I was looking into the tests. But I need a multigpu box to test. So it's taking some time to get running.

albanD commented:
After investigation, the problem is the same as: https://github.com/pytorch/pytorch/pull/13452#issuecomment-439232114 
The Broadcast Function from distributed does bad things.
This PR fix the inplace detection for different models, and the Broadcast Function should definitely not be doing what it does.

vincentqb reviewed (changes requested):
Causes issues in #30334 ? Thanks for identifying this.

albanD commented:
Closing this until we fix the Broadcast issue described in  https://github.com/pytorch/pytorch/pull/13452#issuecomment-439232114
